fcm/fcm.go

```diff
@@ -0,0 +1,200 @@
+	f := FCM{
+		fcmServerBindURL,
+		"",
+		time.Time{},
+		clientID,
+		proxyName,
+		0,
+		FcmSubscribe,
+		notifications,
+		make(chan struct{}),
+		make(chan struct{}),
+	}
+	return &f, nil
+}
+
+//  get token from GCP and connect to FCM.
+func (f *FCM) Init() {
+	iidToken := f.GetToken()
+	if err := f.ConnectToFcm(f.notifications, iidToken, f.dead, f.quit); err != nil {
+		for err != nil {
+			log.Errorf("FCM restart failed, will try again in 10s: %s", err)
+			time.Sleep(10 * time.Second)
+			err = f.ConnectToFcm(f.notifications, iidToken, f.dead, f.quit)
+		}
+		log.Info("FCM conversation restarted successfully")
+	}
+
+	go f.KeepFcmAlive()
+}
+
+// Quit terminates the FCM conversation so that new jobs stop arriving.
+func (f *FCM) Quit() {
+	// Signal to KeepFCMAlive.
+	close(f.quit)
+}
+
+// Fcm notification listener
+func (f *FCM) ConnectToFcm(fcmNotifications chan<- notification.PrinterNotification, iidToken string, dead chan<- struct{}, quit chan<- struct{}) error {
+	log.Debugf("Connecting to %s?token=%s", f.fcmServerBindURL, iidToken)
+	resp, err := http.Get(fmt.Sprintf("%s?token=%s", f.fcmServerBindURL, iidToken))
+	if err != nil {
+		// failed for ever no need to retry
+		quit <- struct{}{}
+		return err
+	}
+	if resp.StatusCode == 200 {
+		reader := bufio.NewReader(resp.Body)
+		go func() {
+			for {
+				line, err := reader.ReadBytes('\n')
+				if err == nil || err == io.EOF {
+					printerId := GetPrinterID(string(line))
+					if printerId != "" {
+						pn := notification.PrinterNotification{printerId, notification.PrinterNewJobs}
+						go func() {
+							fcmNotifications <- pn
+						}()
+					}
+					if err == io.EOF {
+						log.Info("DRAIN message received, client reconnecting.")
+						dead <- struct{}{}
+						break
+					}
+				} else {
+					// stop listening unknown error happened.
+					log.Errorf("%v", err)
+					quit <- struct{}{}
+					break
+				}
+			}
+		}()
+	}
+	return nil
+}
+
+// Restart FCM connection when lost.
+func (f *FCM) KeepFcmAlive() {
+	for {
+		select {
+		case <-f.dead:
+			iidToken := f.GetToken()
+			log.Error("FCM conversation died; restarting")
+			if err := f.ConnectToFcm(f.notifications, iidToken, f.dead, f.quit); err != nil {
+				for err != nil {
+					log.Errorf("FCM connection restart failed, will try again in 10s: %s", err)
+					time.Sleep(10 * time.Second)
+					err = f.ConnectToFcm(f.notifications, iidToken, f.dead, f.quit)
+				}
+				log.Info("FCM conversation restarted successfully")
+			}
+
+		case <-f.quit:
+			log.Info("Fcm client Quitting ...")
+			// quitting keeping alive
+			return
+		}
+	}
+}
+
+// Returns cached token and Refresh token if needed.
+func (f *FCM) GetToken() string {
+	if f.tokenRefreshTime == (time.Time{}) || time.Now().UTC().Sub(f.tokenRefreshTime).Seconds() > f.fcmTTLSecs {
```

Comment:
When would f.tokenRefreshTime == (time.Time{})?

Reply:
when connectors starts it is initialized with time.Time{}, this is updated when new token is received from GCP along with its TTL.